### ISSUE TITLE
Generate Missing Image Sizes

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#4764]
## Feature Description

When using a story poster image that was uploaded prior to the Web Stories plugin being active, that poster image will not be in the image sizes expected by the format (poster / square / landscape). This can also happen for newly uploaded images due to timeouts and the like.

In this case, the plugin can end up using the original image size for the poster, which is not ideal.

We should generate missing poster images instead, to ensure correct AMP output.

**Edit (2020-12-27):** The same goes for the publisher logo, which when not cropped correctly will simply be ignored.

There are two cases to consider here:

1. New story: User selects an existing image as the poster image which does not have the sizes we need.
1. Existing stories: User has already created a story with a poster with incorrect image sizes and we need to update it.

The first case we can solve in the editor when editing/saving/publishing the story.

The second case we can handle in the background, either using our upgrader task or a cron job. Now, image processing is a very intensive task, so this part is better suited for a cron task.

**Existing stories**

First, we're going to add a new database upgrade routine to fetch all stories with a poster image. The list of all poster image IDs is then stored somewhere.

Using WordPress' built-in cron-like scheduler, we can run a background task every X hours/days/weeks (TBD) to go through this list and process the next image in line. The processed image is then removed from the queue.

**New story**

Possibilities:

* Use existing `/web-stories/v1/media/<id>/post-process` REST API endpoint to trigger process.  
    It can be called upon setting the featured image, or even when saving the story (i.e. when editing a story which already has a featured image).
    Upside: separate HTTP request that doesn't affect UX.
* Call `wp_update_image_subsizes()` when saving a story.  
    Downside: expensive process, makes saving slow.
* When saving a story, add the featured image to the cron queue and let the cron job handle this instead.  
   Downside: after publishing, the image size will be incorrect for a period of time until the cron job runs.

## Alternatives Considered

* Not doing anything, keep status quo
* Not adding the poster image if the size is wrong/missing

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by swissspidy** [comment]
@o-fernandez FYI, wrote down some thoughts here on how we could fix cases of incorrect poster image sizes. Please LMK what you think.

**Comment by swissspidy** [comment]
@spacedmonkey Is this an accurate summary of our discussion?

**Comment by spacedmonkey** [comment]
We could call the process endpoint when a story is save for images in page / feature image?

**Comment by swissspidy** [comment]
Yes, that's one of the possibilities mentioned above. They all have pros and cons.

> images in page

So far this issue only covers the featured image, as that's the priority. If we also want to cover images within the story itself, we need to re-evaluate the possibilites a bit.

Post-processing all images would require updating elements using `updateElementsByResourceId`, so this would ideally run upon insertion, not on saving.

However, these probably have less of a priority than the featured image since they don't affect AMP validation.

So... do we want to post-process **_all_** images? I don't think so.

**Comment by spacedmonkey** [comment]
I am not sure about forcing regeneration on save. If we hook into say, `save_post` action, it may mean that save post might timeout. The image processing needs to happen in another process, so another API request. 

When we trigger this request, is really up to you @swissspidy. It is an async request that we don't really care about the result, we can fire off a request whenever we want, with little effect on the editors performance.

**Comment by swissspidy** [comment]
Right, I think doing it on save in PHP is off the table.

That leaves two other possibilites, as stated in the description:

> * Use existing post-process REST API endpoint to trigger process.
>  It can be called upon setting the featured image, or even when saving the story
> * When saving a story, add the featured image to the cron queue and let the cron job handle this instead.
>  Downside: after publishing, the image size will be incorrect for a period of time until the cron job runs.

I suppose using the REST API endpoint is fine.

----

Still the open question from above: do we also want to post-process _all_ images? That changes the scope quite a bit.

**Comment by swissspidy** [comment]
@spacedmonkey Publisher logos should also be post-processed. Something for the cron job? Or to do upon logo selection in the dashboard/editor?

**Comment by spacedmonkey** [comment]
We could call `/wp/v2/media/<id>/post-process` REST API to trigger regeneration of images. 

https://core.trac.wordpress.org/ticket/47987

**Comment by spacedmonkey** [comment]
While working on a personal project over the weekend, I looked into `/wp/v2/media/<id>/post-process` endpoint. To my happy surprise, it is extremely easy to use. So I mocked something up in 10 minutes to backfill image sizes. See the [try/regenerate-sizes](https://github.com/google/web-stories-wp/tree/try/regenerate-sizes). This works like image poster image generation. It just calls the REST API on any image that has a missing size. This work well. One thing is doesn't do currently is update the existing state. Already inserted image in page and state in the library will not be refreshed. But when the editor is reloaded or media library is loaded in anyway, like search it would refresh.
